# Laboratorio #5 — Agentes en el Arcade Learning Environment (ALE): Space Invaders


## 1. Librerías necesarias

In [ ]:
#descomentar si ya se tienen 
# %pip install gymnasium ale-py moviepy imageio imageio-ffmpeg


## 2. Módulo de funciones: `ale_utils.py`

El módulo completo se encuentra en el archivo `ale_utils.py` (al mismo nivel que este archivo)

In [ ]:
print(open('ale_utils.py').read())

"""
Módulo de funciones reutilizables para crear entornos de Gymnasium/ALE,
ejecutar agentes (aleatorio o basado en reglas) y grabar video de las
partidas.
"""

import os
import numpy as np
import gymnasium as gym
import ale_py

# Registrar los entornos de ALE (ALE/SpaceInvaders-v5, etc.) en Gymnasium.
gym.register_envs(ale_py)


def crear_entorno(nombre_entorno, video_folder=None, name_prefix="video",
                   episode_trigger=None, render_mode="rgb_array", **kwargs):
    """
    Crea y retorna un entorno de Gymnasium.

    Parámetros
    ----------
    nombre_entorno : str
        Id del entorno a crear.  "ALE/SpaceInvaders-v5".
    video_folder : str, opcional
        Carpeta donde se guardarán los videos. Si se especifica, el entorno
        se envuelve con gymnasium.wrappers.RecordVideo.
    name_prefix : str
        Prefijo para el nombre de los archivos de video generados.
    episode_trigger : callable, opcional
        Función que recibe el número de episodio (in

## 3. Importar el módulo

In [9]:
from ale_utils import (
    crear_entorno,
    agente_aleatorio,
    agente_regla_simple,
    ejecutar_episodio,
    generar_video_agente,
)

import gymnasium as gym
print("Gymnasium:", gym.__version__)


Gymnasium: 1.2.3


## 4. Validación del módulo con `CartPole-v1`

Para probar que lo realizado en ale_uitls.py funciona con cualquier entorno (se usa el del lab pasado para probar)

In [4]:
env_cartpole = crear_entorno("CartPole-v1")
resultado_cartpole = ejecutar_episodio(env_cartpole, agente_aleatorio, max_steps=500)
env_cartpole.close()

print("Pasos:", resultado_cartpole["pasos"])
print("Recompensa total:", resultado_cartpole["recompensa_total"])


Pasos: 15
Recompensa total: 15.0


## 5. Agente aleatorio en `ALE/SpaceInvaders-v5` (con video)

Generar y guardar un video de 1 episodio completo del agente aleatorio
jugando Space Invaders. También se muestran: número de pasos sobrevividos y la
recompensa total obtenida.

In [5]:
resultado_random = generar_video_agente(
    nombre_entorno="ALE/SpaceInvaders-v5",
    funcion_agente=agente_aleatorio,
    video_folder="videos/agente_aleatorio",
    name_prefix="space_invaders_random",
    n_episodios=1,
)

print("Videos generados:")
for v in resultado_random["videos"]:
    print(" -", v)

print()
print("Métricas por episodio (agente aleatorio):")
for i, m in enumerate(resultado_random["metricas"]):
    print(f"Episodio {i}: pasos={m['pasos']}, recompensa_total={m['recompensa_total']}, "
          f"terminated={m['terminated']}, truncated={m['truncated']}")


C:\Users\darta\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\gymnasium\wrappers\rendering.py:293: UserWarning: WARN: Overwriting existing videos at c:\Users\darta\OneDrive\Desktop\REN\UVG\2026\Segundo ciclo\Deep\lab5_deepLearning\videos\agente_aleatorio folder (try specifying a different `video_folder` for the `RecordVideo` wrapper if this is not desired)
  logger.warn(


Videos generados:
 - videos/agente_aleatorio\space_invaders_random-episode-0.mp4

Métricas por episodio (agente aleatorio):
Episodio 0: pasos=282, recompensa_total=15.0, terminated=True, truncated=False


### Reproducir el video del agente aleatorio en el notebook

In [6]:
from IPython.display import Video

Video(resultado_random["videos"][0], embed=True, width=480)


## 6. Agente basado en regla simple en `ALE/SpaceInvaders-v5` (con video)
Igual que el lab pasado, no solo un random sino también se probó con un con una política simple (sin aprendizaje)

In [7]:
resultado_regla = generar_video_agente(
    nombre_entorno="ALE/SpaceInvaders-v5",
    funcion_agente=agente_regla_simple,
    video_folder="videos/agente_regla_simple",
    name_prefix="space_invaders_regla",
    n_episodios=1,
)

print("Videos generados:")
for v in resultado_regla["videos"]:
    print(" -", v)

print()
print("Métricas por episodio (agente de regla simple):")
for i, m in enumerate(resultado_regla["metricas"]):
    print(f"Episodio {i}: pasos={m['pasos']}, recompensa_total={m['recompensa_total']}, "
          f"terminated={m['terminated']}, truncated={m['truncated']}")


C:\Users\darta\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\gymnasium\wrappers\rendering.py:293: UserWarning: WARN: Overwriting existing videos at c:\Users\darta\OneDrive\Desktop\REN\UVG\2026\Segundo ciclo\Deep\lab5_deepLearning\videos\agente_regla_simple folder (try specifying a different `video_folder` for the `RecordVideo` wrapper if this is not desired)
  logger.warn(


Videos generados:
 - videos/agente_regla_simple\space_invaders_regla-episode-0.mp4

Métricas por episodio (agente de regla simple):
Episodio 0: pasos=545, recompensa_total=155.0, terminated=True, truncated=False


In [8]:
Video(resultado_regla["videos"][0], embed=True, width=480)


## 7. Resumen de resultados

| Agente | Pasos sobrevividos | Recompensa total |
|---|---|---|
| Aleatorio (`agente_aleatorio`) | 282 | 15 |
| Regla simple (`agente_regla_simple`) | 545 | 155 |

(Pueden variar por cálculos aleatorios)

